File: backend/app/services/generation.py
"""
Answer generation service using LLM with RAG
"""

from typing import List, Dict, Any, Tuple
from loguru import logger
import openai
from app.core.config import settings


SYSTEM_PROMPTS = {
    "ar": """أنت محامي، مساعد قانوني ذكي متخصص في القانون المغربي. مهمتك هي:
1. تقديم معلومات قانونية دقيقة ومفيدة
2. الاستشهاد بالمصادر الرسمية (الجريدة الرسمية، القوانين، المراسيم)
3. شرح النصوص القانونية بطريقة مبسطة وواضحة
4. التحدث بالدارجة المغربية عند الحاجة مع الحفاظ على الدقة
5. التنبيه دائماً بأن هذه معلومات تعليمية وليست استشارة قانونية مهنية

قواعد مهمة:
- استخدم فقط المعلومات الموجودة في المصادر المقدمة
- اذكر رقم المادة والقانون في كل إجابة
- إذا لم تجد المعلومة، قل ذلك بوضوح
- لا تخترع معلومات قانونية""",
    "fr": """Tu es Mo7ami, un assistant juridique intelligent spécialisé dans le droit marocain. Ta mission est de:
1. Fournir des informations juridiques précises et utiles
2. Citer les sources officielles (Bulletin Officiel, lois, décrets)
3. Expliquer les textes juridiques de manière simple et claire
4. Parler en français standard tout en restant accessible
5. Toujours rappeler que ce sont des informations éducatives, pas des conseils juridiques professionnels

Règles importantes:
- N'utiliser que les informations présentes dans les sources fournies
- Citer l'article et la loi dans chaque réponse
- Si l'information n'est pas trouvée, le dire clairement
- Ne jamais inventer d'informations juridiques""",
}


async def generate_answer(
    query: str, documents: List[Dict[str, Any]], language: str = "ar"
) -> Tuple[str, List[Dict[str, str]]]:
    """
    Generate answer using LLM with retrieved documents

    Args:
        query: User's question
        documents: Retrieved relevant documents
        language: Response language (ar or fr)

    Returns:
        Tuple of (answer text, list of citations)
    """
    try:
        logger.info(f"Generating answer for query in {language}")

        # Build context from retrieved documents
        context = build_context(documents, language)

        # Build user prompt
        user_prompt = f"""السؤال: {query}

المصادر القانونية المتاحة:
{context}

الرجاء تقديم إجابة دقيقة مع الاستشهاد بالمصادر."""

        if language == "fr":
            user_prompt = f"""Question: {query}

Sources juridiques disponibles:
{context}

Veuillez fournir une réponse précise avec citations."""

        # Call OpenAI API
        client = openai.AsyncOpenAI(api_key=settings.OPENAI_API_KEY)
        response = await client.chat.completions.create(
            model=settings.OPENAI_MODEL,
            messages=[
                {"role": "system", "content": SYSTEM_PROMPTS[language]},
                {"role": "user", "content": user_prompt},
            ],
            temperature=0.3,  # Lower temperature for factual responses
            max_tokens=1500,
        )

        answer = response.choices[0].message.content

        # Extract citations from documents
        citations = extract_citations(documents)

        logger.info("Answer generated successfully")
        return answer, citations

    except Exception as e:
        logger.error(f"Generation error: {e}")
        raise


def build_context(documents: List[Dict[str, Any]], language: str) -> str:
    """Build context string from retrieved documents"""
    context_parts = []

    for i, doc in enumerate(documents, 1):
        source = doc.get("source", "Unknown")
        ref = doc.get("official_ref", "")
        content = doc.get("content", "")

        context_parts.append(f"[{i}] {source} ({ref}):\n{content}\n")

    return "\n".join(context_parts)


def extract_citations(documents: List[Dict[str, Any]]) -> List[Dict[str, str]]:
    """Extract citations from documents"""
    citations = []

    for doc in documents:
        citation = {
            "source": doc.get("source", ""),
            "article": doc.get("article", ""),
            "reference": doc.get("official_ref", ""),
            "url": doc.get("url", ""),
        }
        citations.append(citation)

    return citations

File: lib/rag/generation.ts
/**
 * Answer Generation Service for Mo7ami RAG Pipeline
 * Generates legal answers using retrieved context and OpenAI GPT
 */

import OpenAI from 'openai';
import { retrieveRelevantChunks, detectLegalDomain, type RetrievedChunk } from './retrieval';
import { detectLanguage, type Language } from '../utils/language';

const openai = new OpenAI({
  apiKey: process.env.OPENAI_API_KEY || '',
});

const GPT_MODEL = process.env.OPENAI_MODEL || 'gpt-4o-mini';  // Pinned to stable model (GPT-5 doesn't exist)

export interface Citation {
  source: string;
  article?: string;
  reference: string;
  url?: string;
  similarity?: number;
}

export interface GeneratedAnswer {
  answer: string;
  citations: Citation[];
  language: Language;
  retrievedChunks: number;
  domain: string | null;
}

const SYSTEM_PROMPTS = {
  ar: `أنت محامي، مساعد قانوني ذكي ومتخصص في القانون المغربي. أنت خبير في جميع القوانين المغربية مع معرفة عميقة بالتفاصيل.

قواعد مهمة للأرقام:
- اكتب الأرقام بالحروف: "خمس سنوات" وليس "5 سنوات"
- للمواد القانونية: "المادة خمسمائة وخمسة" وليس "المادة 505"
- للسنوات: "ألف وتسعمائة واثنان وستون" وليس "1962"
- للمبالغ: "مائتان درهم" وليس "200 درهم"

مهمتك - كن مفصلاً ودقيقاً:
- قدم معلومات قانونية شاملة ومفصلة بناءً على **النصوص القانونية المقدمة لك**
- استشهد **فقط** بالمواد والفصول الموجودة في السياق المعطى
- إذا لم يكن السياق كافياً، اذكر ذلك بوضوح
- اذكر أرقام المواد الدقيقة، الفصول، الفقرات
- قدم إرشادات عملية خطوة بخطوة
- استخدم الدارجة المغربية بطريقة احترافية

هيكل الإجابة المطلوب:

1️⃣ **الإجابة المباشرة** (جملتين)

2️⃣ **التفاصيل القانونية:**
   - استشهد بالنصوص القانونية من السياق المعطى
   - اشرح الشروط والإجراءات
   - المدد الزمنية والمبالغ إن وجدت

3️⃣ **الإجراءات العملية:** (إذا كانت مطبقة)

4️⃣ **مثال توضيحي:** (إذا كان مفيداً)

5️⃣ **📚 المصادر القانونية:**
اذكر المصادر بناءً على السياق المعطى بهذا الشكل:
• **[اسم القانون]**
  - المادة/الفصل: [رقم بالحروف]
  - المرجع الرسمي: [ظهير/قانون]
  - الرابط: https://www.sgg.gov.ma

6️⃣ **سؤال المتابعة:**
"واش بغيتي تفاصيل أكثر على شي نقطة معينة؟ 😊"

⚠️ تحذير قانوني (فقط للحالات الشخصية):
"⚖️ ملاحظة: هذه معلومات قانونية عامة للتعليم. لحالتك الخاصة، يُنصح باستشارة محامي مختص."`,

  fr: `Tu es Mo7ami, un assistant juridique intelligent spécialisé dans le droit marocain. Tu es un expert de toutes les lois marocaines avec une connaissance approfondie des détails.

Règles importantes pour les chiffres:
- Écris les nombres en lettres: "cinq ans" et non "5 ans"
- Pour les articles de loi: "l'article cinq cent cinq" et non "l'article 505"
- Pour les années: "mille neuf cent soixante-deux" et non "1962"
- Pour les montants: "deux cents dirhams" et non "200 dirhams"

Ta mission - Sois détaillé et précis:
- Fournis des informations juridiques complètes basées **uniquement sur les textes légaux fournis**
- Cite **seulement** les articles présents dans le contexte donné
- Si le contexte n'est pas suffisant, indique-le clairement
- Mentionne les numéros d'articles précis, chapitres, paragraphes
- Fournis des conseils pratiques étape par étape
- Utilise le français de manière professionnelle

Structure de réponse requise:

1️⃣ **Réponse directe** (deux phrases)

2️⃣ **Détails juridiques:**
   - Cite les textes légaux du contexte fourni
   - Explique les conditions et procédures
   - Délais et montants si applicables

3️⃣ **Procédures pratiques:** (si applicable)

4️⃣ **Exemple concret:** (si utile)

5️⃣ **📚 Sources juridiques:**
Mentionne les sources basées sur le contexte:
• **[Nom de la loi]**
  - Article/Chapitre: [numéro en lettres]
  - Référence officielle: [Dahir/Loi]
  - Lien: https://www.sgg.gov.ma

6️⃣ **Question de suivi:**
"Voulez-vous plus de détails sur un point précis? 😊"

⚖️ Avertissement (uniquement pour cas personnels):
"⚖️ Note: Ces informations juridiques sont générales et à but éducatif. Pour votre cas spécifique, consultez un avocat."`,
};

/**
 * Generate legal answer with RAG (Retrieval-Augmented Generation)
 */
export async function generateLegalAnswer(
  query: string,
  language?: Language
): Promise<GeneratedAnswer> {
  try {
    // Detect language if not provided
    const detectedLanguage = language || detectLanguage(query);

    // Detect legal domain for better filtering
    const domain = detectLegalDomain(query);

    // Retrieve relevant document chunks
    const retrievedChunks = await retrieveRelevantChunks(query, {
      matchThreshold: 0.30,  // Increased for quality matching (was 0.05)
      matchCount: 10,
      filterDomain: domain,
    });

    // If no chunks found, provide a fallback response
    if (retrievedChunks.length === 0) {
      return generateFallbackResponse(query, detectedLanguage, domain);
    }

    // Build context from retrieved chunks
    const context = buildContext(retrievedChunks, detectedLanguage);

    // Generate answer using GPT with context
    const answer = await generateAnswerWithContext(query, context, detectedLanguage);

    // Extract citations from retrieved chunks
    const citations = extractCitations(retrievedChunks);

    return {
      answer,
      citations,
      language: detectedLanguage,
      retrievedChunks: retrievedChunks.length,
      domain: domain || null,
    };
  } catch (error) {
    console.error('Error generating legal answer:', error);
    throw new Error('Failed to generate answer');
  }
}

/**
 * Build context from retrieved chunks
 */
function buildContext(chunks: RetrievedChunk[], language: Language): string {
  const intro =
    language === 'ar'
      ? 'النصوص القانونية المرجعية:\n\n'
      : 'Textes juridiques de référence:\n\n';

  const contextParts = chunks.map((chunk, index) => {
    const article = chunk.articleNumber
      ? language === 'ar'
        ? `المادة ${chunk.articleNumber}`
        : `Article ${chunk.articleNumber}`
      : '';

    return `[${index + 1}] ${article}\n${chunk.content}\n(Similarité: ${(chunk.similarity * 100).toFixed(1)}%)`;
  });

  return intro + contextParts.join('\n\n---\n\n');
}

/**
 * Generate answer using GPT with retrieved context
 */
async function generateAnswerWithContext(
  query: string,
  context: string,
  language: Language
): Promise<string> {
  const systemPrompt = SYSTEM_PROMPTS[language];

  const userPrompt =
    language === 'ar'
      ? `السياق القانوني:\n\n${context}\n\n---\n\nالسؤال: ${query}\n\nأجب بناءً على النصوص القانونية المقدمة فقط. إذا لم يكن السياق كافياً، اذكر ذلك.`
      : `Contexte juridique:\n\n${context}\n\n---\n\nQuestion: ${query}\n\nRéponds en te basant uniquement sur les textes fournis. Si le contexte n'est pas suffisant, indique-le.`;

  const response = await openai.chat.completions.create({
    model: GPT_MODEL,
    messages: [
      { role: 'system', content: systemPrompt },
      { role: 'user', content: userPrompt },
    ],
    temperature: 0.7,
    max_tokens: 1000,
  });

  return response.choices[0].message.content || '';
}

/**
 * Extract citations from retrieved chunks
 */
function extractCitations(chunks: RetrievedChunk[]): Citation[] {
  return chunks.map((chunk) => ({
    source: chunk.metadata?.documentTitle || 'القانون المغربي',
    article: chunk.articleNumber || undefined,
    reference: chunk.metadata?.officialRef || 'مصادر رسمية',
    url: 'https://www.sgg.gov.ma',
    similarity: chunk.similarity,
  }));
}

/**
 * Generate fallback response when no documents are retrieved
 */
async function generateFallbackResponse(
  query: string,
  language: Language,
  domain: string | null | undefined
): Promise<GeneratedAnswer> {
  const fallbackMessage =
    language === 'ar'
      ? `عذراً، لم أجد نصوصاً قانونية محددة في قاعدة البيانات تتطابق مع سؤالك.

يمكنني مساعدتك بمعلومات عامة، لكن أنصحك بـ:
1. إعادة صياغة السؤال بطريقة أكثر تحديداً
2. استشارة محامٍ مختص للحصول على معلومات دقيقة
3. زيارة موقع الأمانة العامة للحكومة: https://www.sgg.gov.ma

واش بغيتي تعيد صياغة السؤال؟`
      : `Désolé, je n'ai pas trouvé de textes juridiques spécifiques dans la base de données correspondant à votre question.

Je peux vous aider avec des informations générales, mais je vous conseille de:
1. Reformuler la question de manière plus précise
2. Consulter un avocat pour des informations exactes
3. Visiter le site du Secrétariat Général du Gouvernement: https://www.sgg.gov.ma

Voulez-vous reformuler votre question?`;

  return {
    answer: fallbackMessage,
    citations: [
      {
        source: language === 'ar' ? 'الأمانة العامة للحكومة' : 'Secrétariat Général du Gouvernement',
        reference: language === 'ar' ? 'مصادر رسمية' : 'Sources officielles',
        url: 'https://www.sgg.gov.ma',
      },
    ],
    language,
    retrievedChunks: 0,
    domain: domain || null,
  };
}
